server/src/com/thoughtworks/go/config/update/PatchEnvironmentCommand.java

```diff
@@ -54,66 +55,80 @@ public PatchEnvironmentCommand(GoConfigService goConfigService, EnvironmentConfi
         }
-        return true;
-
     }
 
-
     @Override
-    public void update(CruiseConfig preprocessedConfig) throws Exception {
-        EnvironmentsConfig environments = preprocessedConfig.getEnvironments();
-        int index = environments.indexOf(environmentConfig);
-        EnvironmentConfig preprocessedEnvironmentConfig = environments.get(index);
-
-        if(isValidConfig(preprocessedConfig)){
-            for (String uuid : agentsToAdd) {
-                preprocessedEnvironmentConfig.addAgent(uuid);
+    public boolean isValid(CruiseConfig preprocessedConfig) {
+        boolean isValid = validateRemovePipelines(preprocessedConfig);
+        isValid = isValid && validateRemoveAgents(preprocessedConfig);
+        return isValid && super.isValid(preprocessedConfig);
+    }
+
+    private boolean validateRemoveAgents(CruiseConfig preprocessedConfig) {
+        EnvironmentConfig preprocessedEnvironmentConfig = preprocessedConfig.getEnvironments().find(environmentConfig.name());
+        if(preprocessedEnvironmentConfig instanceof MergeEnvironmentConfig){
+            for (String agentToRemove : agentsToRemove) {
+                if(preprocessedEnvironmentConfig.containsAgentRemotely(agentToRemove)){
+                    String message = String.format("Can not remove agent with uuid '%s'. As it is defined remotely in environment '%s'", agentToRemove, environmentConfig.name());
+                    result.badRequest(actionFailed.addParam(message));
+                    return false;
+                }
             }
+        }
 
-            for (String uuid : agentsToRemove) {
-                preprocessedEnvironmentConfig.removeAgent(uuid);
+        EnvironmentConfig environmentConfig = this.environmentConfig;
+        for (String agentToRemove : agentsToRemove) {
+            if(!environmentConfig.hasAgent(agentToRemove)){
+                String message = String.format("Agent with uuid '%s' does not exist in environment '%s'", agentToRemove, environmentConfig.name());
+                result.badRequest(actionFailed.addParam(message));
+                return false;
             }
+        }
+
+        return true;
+    }
 
-            for (String pipelineName : pipelinesToAdd) {
-                preprocessedEnvironmentConfig.addPipeline(new CaseInsensitiveString(pipelineName));
+    private boolean validateRemovePipelines(CruiseConfig preprocessedConfig) {
+        EnvironmentConfig preprocessedEnvironmentConfig = preprocessedConfig.getEnvironments().find(environmentConfig.name());
+        if(preprocessedEnvironmentConfig instanceof MergeEnvironmentConfig){
+            for (String pipelineToRemove : pipelinesToRemove) {
+                if(preprocessedEnvironmentConfig.containsPipelineRemotely(new CaseInsensitiveString(pipelineToRemove))){
+                    String message = String.format("Can not remove pipeline '%s'. As it is defined remotely in environment '%s'", pipelineToRemove, environmentConfig.name());
```

Comment:
Would something like `Pipeline '%s' cannot be removed from environment '%s' as the association has been defined remotely in [<getOrigin().displayName()>]` sound better? And something similar for the agent message too.